src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,219 @@ impl HeartbeatHandlerGroup {
+        pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+        sequence: Sequence,
+        check_interval_millis: u64,
+    ) -> MailboxRef {
+        let ins = Arc::new(Self {
+            pushers,
+            sequence,
+            senders: DashMap::default(),
+            recervers: DashMap::default(),
+            timeouts: DashMap::default(),
+        });
+
+        if check_interval_millis > 0 {
+            let timeout_checker = ins.clone();
+            common_runtime::spawn_bg(async move {
+                timeout_checker
+                    .check_timeout_bg(check_interval_millis)
+                    .await;
+            });
+        }
+
+        ins
+    }
+
+    async fn check_timeout_bg(&self, interval_millis: u64) {
+        let mut interval = tokio::time::interval(Duration::from_millis(interval_millis));
+
+        loop {
+            interval.tick().await;
+
+            let now = Duration::from_millis(common_time::util::current_time_millis() as u64);
+            let timeout_ids = self
+                .timeouts
+                .iter()
+                .filter_map(|entry| {
+                    let (id, deadline) = entry.pair();
+                    if deadline < &now {
+                        Some(*id)
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            for id in timeout_ids {
+                let _ = self
+                    .on_recv(id, Err(error::MailboxTimeoutSnafu { id }.build()))
+                    .await;
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Mailbox for HeartbeatMailbox {
+    async fn send(&self, ch: &Channel, mut msg: MailboxMessage) -> Result<MessageId> {
+        let message_id = self.sequence.next().await?;
+
+        msg.id = message_id;
+        let res = HeartbeatResponse {
+            header: None,
+            mailbox_messages: vec![msg],
+        };
+
+        let pusher_id = match ch {
+            Channel::Datanode(id) => format!("0-{}", id),
+            Channel::Frontend(id) => format!("1-{}", id),
+        };
+        let pushers = self.pushers.read().await;
+        let pusher = pushers
+            .get(&pusher_id)
+            .context(error::PusherNotFoundSnafu { pusher_id })?;
+        pusher.send(Ok(res)).await.map_err(|e| {
+            error::PushMessageSnafu {
+                err_msg: e.to_string(),
+            }
+            .build()
+        })?;
+
+        let (tx, rx) = oneshot::channel();
+        self.senders.insert(message_id, Tx(tx));
+        self.recervers.insert(message_id, Rx(rx));
+
+        Ok(message_id)
+    }
+
+    async fn recv(&self, id: MessageId) -> Result<MailboxMessage> {
+        let (_, rx) = self
+            .recervers
+            .remove(&id)
+            .context(error::MailboxNotFoundSnafu { id })?;
+        rx.0.await
+            .map_err(|_| error::MailboxClosedSnafu { id }.build())?
+    }
```

Comment:
Need to remove sender from `self.senders`?

Reply:
`on_recv` will remove the sender, while `recv` is responsible for removing the receiver.